### PR TITLE
Added BT node that combines two trajectories 

### PR DESCRIPTION
**Pull request by DavidSpielman**
- [Added bt node that combines two input trajectories and outputs one trajectory](https://github.com/ros-industrial-consortium/scan_n_plan_workshop/commit/c18dfa987f111f82f420b56d6a4ebb3756c6723d)

### CODE REVIEW

File: snp_application/src/bt/snp_bt_ros_nodes.cpp [hunk 1]
```diff
@@ -536,6 +536,51 @@ BT::NodeStatus ReverseTrajectoryNode::tick()
...
+  if (!second_trajectory)
+  {
+    std::stringstream ss;
+    ss << "Failed to get required input value: '" << second_trajectory.error() << "'";
+    config().blackboard->set(ERROR_MESSAGE_KEY, ss.str());
+    return BT::NodeStatus::FAILURE;
+  }
+
+  trajectory_msgs::msg::JointTrajectory first_input = first_trajectory.value();
+  trajectory_msgs::msg::JointTrajectory second_input = second_trajectory.value();
```
Comment (marip8) [inline snp_application/src/bt/snp_bt_ros_nodes.cpp:568]: Let's swap the names of the inputs and trajectories so it is a little more clear

```suggestion
  BT::Expected<trajectory_msgs::msg::JointTrajectory> first_input =
      getInput<trajectory_msgs::msg::JointTrajectory>(FIRST_TRAJECTORY_INPUT_PORT_KEY);
  BT::Expected<trajectory_msgs::msg::JointTrajectory> second_input =
      getInput<trajectory_msgs::msg::JointTrajectory>(SECOND_TRAJECTORY_INPUT_PORT_KEY);

  if (!first_input)
  {
    std::stringstream ss;
    ss << "Failed to get required input value: '" << first_input.error() << "'";
    config().blackboard->set(ERROR_MESSAGE_KEY, ss.str());
    return BT::NodeStatus::FAILURE;
  }

  if (!second_input)
  {
    std::stringstream ss;
    ss << "Failed to get required input value: '" << second_input.error() << "'";
    config().blackboard->set(ERROR_MESSAGE_KEY, ss.str());
    return BT::NodeStatus::FAILURE;
  }

  trajectory_msgs::msg::JointTrajectory first_trajectory = first_input.value();
  trajectory_msgs::msg::JointTrajectory second_trajectory = second_input.value();
```

**Comment by marip8** [comment]
> It seems that while the motion plan is successfully generated, the follow joint trajectory action is unable to process the goal during motion execution because the time between the last point of one trajectory and the first point of the next trajectory is the same (not strictly increasing).

Are you testing this node by itself, or with a trajectory generated by the reverse trajectory block? If you're testing with a reversed trajectory, then it may be the case that the reverse trajectory function has a bug

**Comment by DavidSpielman** [comment]
> > It seems that while the motion plan is successfully generated, the follow joint trajectory action is unable to process the goal during motion execution because the time between the last point of one trajectory and the first point of the next trajectory is the same (not strictly increasing).
> 
> Are you testing this node by itself, or with a trajectory generated by the reverse trajectory block? If you're testing with a reversed trajectory, then it may be the case that the reverse trajectory function has a bug

I am testing this with the node by itself.

**Comment by DavidSpielman** [comment]
I think this bug occurs within the combine method within snp_bt_ros_nodes.cpp. When combining two trajectories, the first point of the second trajectory has a time_from_start value of 0 nanoseconds and 0 seconds. When this time from start value is added to the duration of the first trajectory, the last point of the first trajectory and the first point of the second trajectory have the same values in the time_from_start field (which the follow joint trajectory does not accept). 

As a temporary fix, I have added one nanosecond to the time_from_start field for the first point of the second trajectory, but this is not a robust solution. How should we handle this situation?

**Comment by DavidSpielman** [comment]
The bug has been addressed by not including the first point of the second trajectory when combining trajectories. This assumes that the given trajectories overlap and may need to be adjusted in the future.

File: snp_application/src/bt/snp_bt_ros_nodes.cpp [hunk 2]
```diff
@@ -271,6 +271,11 @@ trajectory_msgs::msg::JointTrajectory combine(const trajectory_msgs::msg::JointT
         first.points.empty() ? builtin_interfaces::msg::Duration() : first.points.back().time_from_start;
     for (const auto& pt : second.points)
     {
+      if (pt == second.points.front())
+      {
+        continue;
+      }
```
Comment (marip8) [inline snp_application/src/bt/snp_bt_ros_nodes.cpp:277]:
```suggestion
    for (std::size_t i = 1; i < second.points.size(); ++i)
    {
      const trajectory_msgs::JointTrajectoryPoint& pt = second.points[i];
```

File: snp_application/src/bt/snp_bt_ros_nodes.cpp [hunk 3]
```diff
@@ -302,6 +307,11 @@ trajectory_msgs::msg::JointTrajectory combine(const trajectory_msgs::msg::JointT
         new_pt.positions[idx] = pt.positions[i];
       }
 
+      if (pt == second.points.front())
+      {
+        continue;
+      }
+
```
Comment (marip8) [inline snp_application/src/bt/snp_bt_ros_nodes.cpp:314]: Let's do this above instead:

```
for (std::size_t i = 1; i < second.points.size(); ++i)
{
  const trajectory_msgs::msg::JointTrajectoryPoint& pt = second.points[i];
```

File: snp_application/src/bt/snp_bt_ros_nodes.cpp [hunk 4]
```diff
@@ -338,6 +348,11 @@ trajectory_msgs::msg::JointTrajectory combine(const trajectory_msgs::msg::JointT
         new_pt.positions[idx] = it->positions[i];
       }
 
+      if (new_pt == second.points.front())
+      {
+        continue;
+      }
+
```
Comment (marip8) [inline snp_application/src/bt/snp_bt_ros_nodes.cpp:355]: Let's do this above instead:

```
for (auto it = first.points.rbegin() + 1; it != first.points.rend(); ++it)
```

**Review by marip8 — CHANGES_REQUESTED** [review]

**Review by marip8 — APPROVED** [review]

**Pull request merged**